Rework "Accept" header handling for HTTP 500 response body on unhandled exceptions

yrodiere opened the pull request:
Fixes #20198

yrodiere committed:
Rework "Accept" header handling for HTTP 500 response body on unhandled exceptions

yrodiere committed:
Fix invalid assertions in vert.x route tests

We were asserting that the error page contains the string "null", in
order to check that an NPE is reported.

But the error page contains javascript code, and that code contains
the javascript keyword "null".

So those assertions were always passing...
In some cases they were passing even though there was no NPE; the
assertion was incorrect and we didn't notice that because it was passing
anyway.

dimas-b committed:
Add HTTP config parameter to control the content type of unexpected error responses

Applications do not always control the context of unexpected
exceptions and may not be able to provide exception mappers in
all cases.

QuarkusErrorHandler previously replied on the `Accept` HTTP header
to decide whether to format the error response as JSON or HTML.

RFC6839 defines the `+json` media type suffix that may be used
by applications to create custom media types compatible with
the JSON format. For example: `application/my-app+json`

In those cases QuarkusErrorHandler cannot reasonably be able to
produce the exact expected response contents type because its
specification is application-dependent. However, it may still be
preferable to return a generic `applicaiton/json` response as
opposed to HTML in those cases.

The new (optional) configuration setting allows applications to
request a specific contents type for formatting unexpected error
responses.

The default Quarkus behaviour remains unchanged. QuarkusErrorHandler
will use the `Accept` header for determining the response contents
type.

cescoffier commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/QuarkusErrorHandler.java:
```diff
@@ -118,28 +134,49 @@ public void accept(Throwable throwable) {
+        event.response().headers().set(HttpHeaderNames.CONTENT_TYPE, contentType + "; charset=utf-8");
+        String escapedDetails = escapeJsonString(details);
+        String escapedStack = escapeJsonString(stack);
+        StringBuilder jsonPayload = new StringBuilder("{\"details\":\"")
```
Unrelated to this PR - I'm wondering if at some point we shouldn't use something like https://github.com/zalando/problem.

cescoffier commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/QuarkusErrorHandler.java:
```diff
@@ -200,4 +237,45 @@ static String escapeJsonString(final String text) {
+        private static final Set<String> SUPPORTED_SET = Set.of(APPLICATION_JSON, TEXT_JSON, TEXT_HTML, APPLICATION_XML,
+                TEXT_XML, APPLICATION_XHTML);
+
+        static String pickFirstSupportedAndAcceptedContentType(HttpServerRequest request) {
```
I'm wondering if Vert.x does not have this code somewhere.

cescoffier replied:
Should we reuse: https://github.com/vert-x3/vertx-web/blob/36bf548a7362336a9838e8260c40b543fc032cc6/vertx-web/src/main/java/io/vertx/ext/web/impl/HeaderParser.java ?

yrodiere replied:
Thanks! I was looking for something like that, but didn't find it. I'll change the code to use it.

yrodiere committed:
Consider the configured unhandled exception content type as a default

... rather than an override.

yrodiere commented:
@cescoffier Thanks for the review. I addressed your comment, and also incorporated #20251, along with an extra commit to change the setting from an override to just a default (see [my comment on that PR](https://github.com/quarkusio/quarkus/pull/20251#discussion_r711906466)).

cescoffier reviewed (approved):
LGTM! Thanks for integrating the _conflicting_ PR.

yrodiere committed:
Change the default format of responses returned by QuarkusErrorHandler

yrodiere committed:
Use consistent phrasing for HTTP 500 error message with or without stack trace

Otherwise it's hard to write tests that work in both native mode
(prod profile) and non-native mode (dev/test profile).
